Launchpad: "Write your first post" task links to first post if you have one

renancarvalho opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/94086
Fixes https://github.com/Automattic/wp-calypso/issues/94086

## Proposed Changes

The "publish your first post" task in launchpad always uses the URL returned by the server.
In combination with https://github.com/Automattic/jetpack/pull/39259, this means sometimes clicking the task takes the user to a draft already in progress.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

As reported in #94086, the user can find themselves in an onboarding flow loop: they draft a first post, head back to the launchpad, click "write your first post", draft a new post, head back to the launchpad, etc. etc.
The user should be more successful completing the task if they don't end up in a loop.

This PR removes some logic that meant the `start-writing` and `design-first` flows were forced to use the `wp-admin` version of the editor. That behaviour still exists but it's now on the backend.

This PR has also retained the behaviour of only appending the `origin` query param for the `start-writing` and `design-first` flows. It's not clear what this query param is used for, but it seems best not to change it if I don't understand it.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Add the Jetpack changes to your sandbox
  * Sandbox `public-api`
  * Run `bin/jetpack-downloader test jetpack-mu-wpcom-plugin feat/launchpad_draft_post`
  * See PCYsg-Osp-p2 for more details
* Build Calypso changes locally
* Test `write` flow with a draft post
  * Create new site using `/start`
  * When asked about goals choose **Write & Publish**
  * Continue on flow and choose **Draft your first post: Start writing** button
  * Confirm the suggested writing ideas modal appears when the editor loads
    * This is controlled by the `new_prompt=true` query param
  * Save your post as draft but *do not publish*
  * Preview the draft in a new tab
  * In the new tab click **Next steps**
  * You are taken to launchpad, at this point you'll need to change the page domain back to `calypso.localhost:3000`
  * Confirm the **Continue to write your first post** button links back to the editor
    * It should re-open your existing post
    * It should use the Calypso editor
    * There is no `new_prompt` modal when the editor loads
  * Now test in `wp-admin` mode
    * Jump straight to `/settings/general/{{ site slug }}` and change the admin interface to **classic**
    * Click **My Home** in the sidebar to get redirected to the launchpad (confirm your still looking at `calypso.localhost:3000`)
    * Confirm the **Continue to write your first post** button links back to the editor
      * It should re-open your existing post
      * It should use the`wp-admin` editor
      * There is no `new_prompt` modal when the editor loads
* Test `write` flow with no draft post in Calypso site
  * Create new site using `/start`
  * When asked about goals choose **Write & Publish**
  * Continue on flow and choose **Draft your first post: Start writing** button
  * Confirm the suggested writing ideas modal appears when the editor loads
    * This is controlled by the `new_prompt=true` query param
  * *do not save a draft*, leave the editor with the (W) button
  * You are taken to launchpad, at this point you'll need to change the page domain back to `calypso.localhost:3000`
  * Confirm the **Write your first post** button links back to the editor
    * It opens a blank editor
    * It should use the Calypso editor
    * This is a `new_prompt` modal when the editor loads (they might have accidentally dropped out of the flow earlier so I think it's reasonable to show it to them here)
  * Now test in `wp-admin` mode
    * Jump straight to `/settings/general/{{ site slug }}` and change the admin interface to **classic**
    * Click **My Home** in the sidebar to get redirected to the launchpad (confirm your still looking at `calypso.localhost:3000`)
    * Confirm the **Continue to write your first post** button links back to the editor
      * It should open a blank editor
      * It should use the`wp-admin` editor
      * There is a `new_prompt` modal when the editor loads
* Test `start-writing` or `design-first` flow
  * Create a new site using `/setup/start-writing`
  * You are eventually taken to the launchpad
  * Confirm the **Write your first post** button links to the editor
    * It should open a blank editor
    * It should use the `wp-admin` editor (remember these flows are special cased to use `wp-admin`)
    * There is no `new_prompt` modal
    * An `origin` query param has been added
  * Save and preview a draft post
  * In the new tab click **Next steps**
  * Back in the launchpad (confirm you're on `calypso.localhost:3000`) check the **Continue to write your first post** button
    * It should open the existing draft
    * It should use the `wp-admin` editor (remember these flows are special cased to use `wp-admin`)
    * There is no `new_prompt` modal
    * An `origin` query param has been added


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] !Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)!
- [ ] ~Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?~
- [ ] ~For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?~

renancarvalho committed:
Use API url

p-jackson commented:
I was trying to figure out why the `calypso_path` gets overridden at all. Why wouldn't it always use the path provided by the API?

I don't know exactly why, but the logic for doing something custom when `isBlogOnboardingFlow` is true has existed ever since the task was first defined in #86490. I think it must be because `isBlogOnboardingFlow` has some client-side state that the Jetpack API can't know about and so the customisation can't be done on the server.

p-jackson commented:
I've been digging to the bottom of why the front end is overriding the `calypso_path` from the server in the first place.

It appears to be because of the special `start-writing` and `design-first` flows that want to default to the `wp-admin` interface. Despite not having `wp-admin` set as their preferred admin interface 🤔 

Having gone through the `start-writing` flow, it has the same UX issue as the regular signup flow (I don't think `start-writing` is connected to the regular signup flow, but you can access it directly at https://wordpress.com/setup/start-writing). Ideally in these special flows it would also open an existing draft if one has already been created.

I think the answer is to move the logic for all flows to the server. The server can be hardcoded to use `wp-admin` if the user is one of the special flows (you can look it up with `get_option( 'site_intent' )`. The client then won't have anything to do except append the `?origin` query param.

---

A follow up bottle pick up task might be to figure out why `wp-admin` isn't set as the default admin interface when sites are created using `start-writing` or `design-first`. If that's what the intention was.

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/post/index.tsx:
```diff
@@ -14,7 +14,7 @@ export const getFirstPostPublished: TaskAction = ( task, flow, context ): Task =
 			( task.completed && isBlogOnboardingFlow( flow || null ) ) ||
 			false,
 		calypso_path: ! isBlogOnboardingFlow( flow || null )
```
It feels like the logic to add the `origin` URL parameter could be done by building on the incoming `task.calypso_path` value.
```suggestion
		calypso_path: calypso_path: addQueryArgs( task.calypso_path, { origin: window.location.origin } );
```

The `origin` URL parameter should be ignored in most cases (though we should double-check!), and we can remove some conditional logic that is really unnecessary.

p-jackson replied:
I did a smoke test to see whether the writing flow works fine when there's an `origin` URL. It seems to ignore it, but it's hard to know what affect it might have with more extensive use. I'm not exactly sure what `origin` does, although I'm guessing it's something to do with making a flow work nicely regardless of whether it's a prod or dev environment.

I'm inclined to leave `origin` logic as it is for now (that is, only include it for some flows).
1. I can't be truly sure that it won't impact anything
2. It makes the URL a lot uglier. Not that the URL is super pretty without it, but it it includes an escaped `://` sequence

daledupreez replied:
FYI, the `origin` URL parameter is used in the following code:
https://github.com/Automattic/wp-calypso/blob/23e5d3ff396834274865eabc42e2cd1454547fa8/apps/wpcom-block-editor/src/wpcom/features/redirect-onboarding-user-after-publishing-post.js#L42-L66

We're discussing whether that component should be removed entirely in https://github.com/Automattic/wp-calypso/issues/94072 - I think the number of cases where it appears is relatively limited (though we might need some tracking to confirm that).

p-jackson replied:
I was worried that with groundskeeping going on at the same time as cleaning up flows that we'd end up in a situation where we feel unsure about whether some coding effort will be worth it 😄

I'm inclined to just fix the product as it currently is today, but then that'd leave us having to come back and removing the `origin` param as part of #94072

Commit:
All flows using first_post_published path returned by server

The first_post_published task was using the calypso_path returned by the
server, expect for `starting-writing` and `design-first` flows, which
wanted to always open the wp-admin editor.
In Automattic/jetpack#39259 the server now always returns a wp-admin
path for these flows, so the special casing is no longer required.

Commit:
Adjust tests to match new first_post_published behaviour

p-jackson commented on client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/post/test/index.ts:
```diff
@@ -16,12 +16,12 @@ const buildContext = ( options?: Partial< TaskContext > ) => {
 
-	it( 'returns the plans page', () => {
+	it( 'returns the calypso_path provided by the server', () => {
 		const context = buildContext( { siteSlug: 'site.wordpress.com' } );
 
 		expect( getFirstPostPublished( task, 'flowId', context ) ).toMatchObject( {
 			useCalypsoPath: true,
-			calypso_path: '/post/site.wordpress.com',
+			calypso_path: 'some-path',
 		} );
 	} );
```
Updating these unit tests feels a bit arbitrary. I looked back in history and the tests were bulk added in https://github.com/Automattic/wp-calypso/pull/86913. They may have been written in such a way that they simply documented the current behaviour.

e.g. this tests name was "returns the plans page", which isn't what it's testing.

Alternatively, now that the client is using the `calypso_path` returned by the server, maybe we should just delete this unit test.

p-jackson commented:
I've re-written the PR description to include various test cases. I believe this should be good for a review now.

arthur791004 reviewed (approved):
Approved! But we have to wait for https://github.com/Automattic/jetpack/pull/39259
